Add Clash.Tutorial content to user-guide (#10)
clash-lang/docs.clash-lang.org: compiler-user-guide/src/developing-hardware/primitives.md

# User-defined primitives
There are times when you already have an existing piece of IP, or there are times where you need the HDL to have a specific shape so that the HDL synthesis tool can infer a specific component.

> In these specific cases you can resort to defining your own HDL primitives. Actually, most of the primitives in Clash are specified in the same way as you will read about in this section.

Reviewer: ```suggestion
In these specific cases you can resort to defining your own HDL primitives.
Actually, most of the primitives in Clash are specified in the same way as you will read about in this section.
```

          if ~ARG[7] ~IF ~ISACTIVEENABLE[4] ~THEN and ~ARG[4] ~ELSE ~FI then
            ~SYM[2](~SYM[5]) <= ~TOBV[~ARG[9]][~TYP[9]];
          end if;
          ~RESULT <= fromSLV(~SYM[2](~SYM[4]))
          -- pragma translate_off
          after 1 ps
          -- pragma translate_on
          ;
        end if;
      end process; ~ELSE
      ~SYM[6] : process(~ARG[3])
      begin
        if ~IF~ACTIVEEDGE[Rising][0]~THENrising_edge~ELSEfalling_edge~FI(~ARG[3]) then
          if ~ARG[7] ~IF ~ISACTIVEENABLE[4] ~THEN and ~ARG[4] ~ELSE ~FI then
            ~SYM[2](~SYM[5]) <= ~ARG[9];
          end if;
          ~RESULT <= ~SYM[2](~SYM[4])
          -- pragma translate_off
          after 1 ps
          -- pragma translate_on
          ;
        end if;
      end process; ~FI
    end block;
    --end blockRam
```

Again, the `name` of the primitive is the fully qualified name of the function you are creating the primitive for.
Because we are creating a _declaration_ primitive the _kind_ must be set to `Declaration`.
Instead of discussing what the individual template holes mean in the above context, we will instead just give a general listing of the available template holes:

> * `~RESULT`: Signal to which the result of a primitive must be assigned
>   to. NB: Only used in a _declaration_ primitive.

Reviewer: ```suggestion
* `~RESULT`: Signal to which the result of a primitive must be assigned to.
  NB: Only used in a _declaration_ primitive.
```

* `~ARG[N]`: `(N+1)`'th argument to the function.

> * `~CONST[N]`: `(N+1)`'th argument to the function. Like `~ARG`, but Clash will try to reduce this to a literal, even if it would otherwise consider it too expensive.
>   And if Clash fails to reduce this argument to a literal it will produce an error.

Reviewer: ```suggestion
* `~CONST[N]`: `(N+1)`'th argument to the function.
  Like `~ARG`, but Clash will try to reduce this to a literal, even if it would otherwise consider it too expensive.
  And if Clash fails to reduce this argument to a literal it will produce an error.
```

> * `~LIT[N]`: `(N+1)`'th argument to the function. Like `~CONST~` but values are   rendered as a bare literals, without any size or type annotations.

Reviewer: ```suggestion
* `~LIT[N]`: `(N+1)`'th argument to the function.
  Like `~CONST` but values are rendered as bare literals, without any size or type annotations.
```

There is a typo in the Haddock: the `@~CONST~` was in my opinion intended to be `@~CONST@`.

  This only works for numeric types, and not for BitVector.
* `~TYP[N]`: VHDL type of the `(N+1)`'th argument.
* `~TYPO`: VHDL type of the result.

> * `~TYPM[N]`: VHDL type_name_ of the `(N+1)`'th argument; used in _type qualification_.
> * `~TYPMO`: VHDL type_name_ of the result; used in _type qualification_.

Reviewer: ```suggestion
* `~TYPM[N]`: VHDL type _name_ of the `(N+1)`'th argument; used in _type qualification_.
* `~TYPMO`: VHDL type _name_ of the result; used in _type qualification_.
```

There we go! When I wrote earlier that Markdown processors disagree on the semantics of an underscore in the middle of a word, I had no idea we were going to run into that problem a little further down already. mdBook doesn't render it as italics.

However, here it's a mistake that it's written as one word anyway.

* `~SIGDO[<HOLE>]`: Create a signal declaration, using `<HOLE>` as the name
  of the signal, and the type of the result.
* `~TYPEL[<HOLE>]`: The element type of the vector type represented by `<HOLE>`.
  The content of `<HOLE>` must either be: `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
* `~COMPNAME`: The name of the component in which the primitive is instantiated.
* `~LENGTH[<HOLE>]`: The vector length of the type represented by `<HOLE>`.
* `~DEPTH[<HOLE>]`: The tree depth of the type represented by `<HOLE>`.
  The content of `<HOLE>` must either be: `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
* `~SIZE[<HOLE>]`: The number of bits needed to encode the type represented by `<HOLE>`.
  The content of `<HOLE>` must either be: `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
* `~IF <CONDITION> ~THEN <THEN> ~ELSE <ELSE> ~FI`: renders the `<ELSE>`
  part when `<CONDITION>` evaluates to _0_, and renders the `<THEN>` in all
  other cases.
  Valid `<CONDITION>`s are `~LENGTH[<HOLE>]`, `~SIZE[<HOLE>]`, `~CMPLE[<HOLE1>][<HOLE2>]`, `~DEPTH[<HOLE>]`, `~VIVADO`, `~IW64`, `~ISLIT[N]`, `~ISVAR[N]`, `~ISACTIVEENABLE[N]`, `~ISSYNC[N]`, and `~AND[<HOLE1>,<HOLE2>,..]`.

> * `~VIVADO`: _1_ when Clash compiler is invoked with the `-fclash-xilinx` or
>   `-fclash-vivado` flag.

Reviewer: ```suggestion
* `~VIVADO`: _1_ when Clash compiler is invoked with the `-fclash-hdlsyn Vivado` (or `Xilinx` or `ISE`) flag.
```

If we're now also fixing bugs in the same commit, as you've done with some earlier details, I have another one!

* `~FROMBV[<HOLE>][<TYPE>]`: create conversion code that so that the expression in `<HOLE>`, which has a bit vector (`std_logic_vector`) type, is converted to type indicated by `<TYPE>`.
  The `<TYPE>` hole must be either `~TYP[N]`, `~TYPO`, or `~TYPEL[<HOLE>]`.
* `~INCLUDENAME[N]`: the generated name of the `N`'th included component.
* `~FILE[<HOLE>]`: The argument mentioned in `<HOLE>` is a file which must be copied to the location of the generated HDL.
* `~GENERATE`: Verilog: create a _generate_ statement, except when already in a _generate_ context.
* `~ENDGENERATE`: Verilog: create an _endgenerate_ statement, except when already in a _generate_ context.
* `~ISLIT[N]`: Is the `(N+1)`'th argument to the function a literal.
* `~ISVAR[N]`: Is the `(N+1)`'th argument to the function explicitly not a literal.
* `~ISSCALAR[N]`: Is the `(N+1)`'th argument to the function a scalar.
  Note that this means different things for different HDLs.
  In (System)Verilog only `Bit` and `Bool` are considered scalar.
  In VHDL, in addition to those two, enumeration types and integers are considered scalar.
* `~TAG[N]`: Name of given domain.

>   Errors when called on an argument which is not a 'KnownDomain', 'Reset', or 'Clock'.

Reviewer: ```suggestion
  Errors when called on an argument which is not a `KnownDomain`, `Reset`, or `Clock`.
```

* `~PERIOD[N]`: Clock period of given domain.

>   Errors when called on an argument which is not a 'Clock', 'Reset', 'KnownDomain' or 'KnownConf'.

Reviewer: ```suggestion
  Errors when called on an argument which is not a `Clock`, `Reset`, `KnownDomain` or `KnownConf`.
```

> * `~ISACTIVEENABLE[N]`: Is the `(N+1)`'th argument a an Enable line __not__ set to a constant True.

Reviewer: ```suggestion
* `~ISACTIVEENABLE[N]`: Is the `(N+1)`'th argument an Enable line __not__ set to a constant True.
```

>   Can be used instead of deprecated (and removed) template tag

Reviewer: We really properly removed that template tag, didn't we? We even removed it here.

* `~ISSYNC[N]`: Does synthesis domain at the `(N+1)`'th argument have synchronous resets.

>   Errors when called on an argument which is not a 'Reset', 'Clock', 'Enable', 'KnownDomain' or 'KnownConf'.

Reviewer: ```suggestion
  Errors when called on an argument which is not a `Reset`, `Clock`, `Enable`, `KnownDomain` or `KnownConf`.
```

* `~ISINITDEFINED[N]`: Does synthesis domain at the `(N+1)`'th argument have defined initial values.

>   Errors when called on an argument which is not a 'Clock', 'Reset', 'Enable', 'KnownDomain' or 'KnownConf'.

Reviewer: ```suggestion
  Errors when called on an argument which is not a `Reset`, `Clock`, `Enable`, `KnownDomain` or `KnownConf`.
```
